crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -607,31 +582,74 @@ where
-                "Consensus engine reached max block."
-            );
-            true
-        } else {
-            false
-        }
-    }
-
     /// Retrieve the block number for the given block hash.
     fn get_block_number(&self, hash: H256) -> Result<Option<BlockNumber>, reth_interfaces::Error> {
         Ok(self.db.view(|tx| tx.get::<tables::HeaderNumbers>(hash))??)
     }
+
+    /// Event handler for events emitted by the [EngineSyncController].
+    ///
+    /// This returns a result to indicate whether the engine future should resolve (fatal error).
+    fn on_sync_event(
+        &mut self,
+        ev: EngineSyncEvent,
+        current_state: &ForkchoiceState,
+    ) -> Option<Result<(), BeaconEngineError>> {
+        match ev {
+            EngineSyncEvent::FetchedFullBlock(block) => {
+                // it is guaranteed that the pipeline is not active at this point.
+
+                // TODO(mattsse): better error handling and start closing the gap if there's any by
+                //  closing the gap either via pipeline, or by fetching the blocks via block number
+                //  [head..FCU.number]
+
+                let hash = block.hash;
+                if self.on_new_payload(block.into()).is_syncing() {
+                    // we're still syncing here, which means the block was not the next canonical
+                    // block so we're starting the pipleine to block instead
+                    self.sync.set_pipeline_sync_target(hash);
+                }
+            }
+            EngineSyncEvent::PipelineStarted(target) => {
+                trace!(target: "consensus::engine", ?target, continuous = target.is_none(), "Started the pipeline");
+                self.metrics.pipeline_runs.increment(1);
+            }
+            EngineSyncEvent::PipelineTaskDropped => {
+                error!(target: "consensus::engine", "Failed to receive spawned pipeline");
+                return Some(Err(BeaconEngineError::PipelineChannelClosed))
+            }
+            EngineSyncEvent::PipelineFinished { result, reached_max_block } => {
+                match result {
+                    Ok(ctrl) => {
+                        if ctrl.is_unwind() {
+                            self.sync.set_pipeline_sync_target(current_state.head_block_hash);
+                        } else if reached_max_block {
+                            // Terminate the sync early if it's reached the maximum user
+                            // configured block.
+                            return Some(Ok(()))
+                        }
+
+                        // Update the state and hashes of the blockchain tree if possible
+                        if let Err(error) = self.restore_tree_if_possible(*current_state) {
+                            error!(target: "consensus::engine", ?error, "Error restoring blockchain tree");
+                            return Some(Err(error.into()))
+                        }
```

Comment:
previously, the pipeline was immediately advanced if the next target is set. what happens now?
https://github.com/paradigmxyz/reth/blob/main/crates/consensus/beacon/src/engine/mod.rs#L724-L731

Reply:
yes but this logic is now part of the sync controller's poll function